csharp/src/Apache.Arrow/Arrays/BinaryArray.cs

```diff
@@ -73,61 +80,88 @@ public TArray Build(MemoryAllocator allocator = default)
             {
                 ValueOffsets.Append(Offset);
 
-                var data = new ArrayData(DataType, ValueOffsets.Length - 1, 0, 0,
-                    new[] { ArrowBuffer.Empty, ValueOffsets.Build(allocator), ValueBuffer.Build(allocator) });
+                var data = new ArrayData(DataType, ValueOffsets.Length - 1, NullCount, 0,
+                    new[] { ValidityBuffer.Build(allocator).ValueBuffer, ValueOffsets.Build(allocator), ValueBuffer.Build(allocator) });
 
                 return Build(data);
             }
 
-            public TBuilder Append(byte value)
+            public TBuilder AppendNull()
+            {
+                ValueOffsets.Append(Offset);
+                Offset++;
+                NullCount++;
+                return Instance;
+            }
+
+            public TBuilder Append(T value)
             {
                 ValueOffsets.Append(Offset);
                 ValueBuffer.Append(value);
                 Offset++;
+                ValidityBuffer.Append(true);
                 return Instance;
             }
 
-            public TBuilder Append(ReadOnlySpan<byte> span)
+            public TBuilder Append(ReadOnlySpan<T> span)
             {
                 ValueOffsets.Append(Offset);
                 ValueBuffer.Append(span);
-                Offset += span.Length;
+                ValidityBuffer.Append(true);
+                Offset += span == Span<T>.Empty
+                                            ? 1
```

Comment:
I'm not sure this is correct. If I append an empty value to the builder, I would expect the `ValueOffsets` to get the current `Offset`, but since there were no values append, `Offset` stays the same. That way the "Length" of the current element will be `0`.

Reply:
Check out https://arrow.apache.org/docs/format/Columnar.html#variable-size-binary-layout for an explanation of how this works. This is the key part:

For example, the position and length of slot j is computed as:

```
slot_position = offsets[j]
slot_length = offsets[j + 1] - offsets[j]  // (for 0 <= j < length)
```

If you add to `Offset` here, that will make this element appear to have `length == 1`, which it shouldn't.

Reply:
Fixed